### ISSUE TITLE
Double underscores in schema field names causing validator some grief

### ISSUE THREAD

**Issue by himynameisdave** [graphql/graphql-js#1085]
_Originally filed in [`eslint-plugin-graphql`](https://github.com/apollographql/eslint-plugin-graphql) over here: https://github.com/apollographql/eslint-plugin-graphql/issues/92_

We have a few fields in our schema that start with a double underscore. One such example is a `__cost` field, which we use to enforce query limits. This causes an error with regards to it being named with a double underscore, and there doesn't appear to be any validation setting that can be passed in to turn this off.

**Error message:**
```
Name "__cost" must not begin with "__", which is reserved by GraphQL introspection. In a future release of graphql this will become a hard error.
```

**Relevant slice of our schema:**
```
...
{
    "name": "__cost",
    "description": null,
    "args": [],
    "type": {
        "kind": "SCALAR",
        "name": "JSONObject",
        "ofType": null
    },
    "isDeprecated": false,
    "deprecationReason": null
}
...
```

Would be nice to either turn this off or be able to pass a validation flag or something. It appears that the function where this error is originating from [already accepts an `isIntrospection`](https://github.com/graphql/graphql-js/blob/6e7d51d6c8fb9ed66d93510fa49477a05dd751af/src/utilities/assertValidName.js#L29) param that would disable this warning. However, [it does not appear that this is ever used](https://github.com/graphql/graphql-js/blob/4f5e351c2afb28e4ae3a44eae76d45d895ff3aa2/src/type/directives.js#L59). 

**TLDR;** allow us to pass that `isIntrospection` var, allowing us to mute that "no underscores in field names" error when validating an introspection query, such as the one required by `eslint-plugin-graphql`.

**Comment by leebyron** [comment]
That `assertValidName` utility can be used externally, and is used when defining the introspection system itself, hence allowing the var.

Unfortunately this warning is likely to become louder instead of configurable in the future as the spec disallows fields named with two underscores to preserve the namespace for introspection features. For example a "__cost" field might represent some future introspection feature which measures the server cpu cost of running a field. The double underscore for introspection features ensures they won't ever conflict with fields in user defined schema.
